tests/core/toolkit/test_sync_to_async.py

```diff
@@ -0,0 +1,131 @@
+
+## Helpers #####################################################################
+
+
+class IterableHelper:
+    def __init__(self, iterable, delay=0, raise_element=None):
+        self.iterable = iterable
+        self.delay = delay
+        self.raise_element = raise_element
+        self.idx = 0
+        self.iter = iter(self.iterable)
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        if self.delay:
+            time.sleep(self.delay)
+        next_element = next(self.iter)
+        if self.idx == self.raise_element:
+            raise RuntimeError("Raising")
+        self.idx += 1
+        return next_element
+
+
+async def azip(*aiterables):
+    aiterators = [aiterable.__aiter__() for aiterable in aiterables]
+    while aiterators:
+        results = await asyncio.gather(
+            *[ait.__anext__() for ait in aiterators],
+            return_exceptions=True,
+        )
+        if any(isinstance(result, StopAsyncIteration) for result in results):
+            return
+        yield tuple(results)
+
+
+## Tests #######################################################################
+
+
+@pytest.mark.asyncio
+async def test_async_wrap_iter_happy():
+    """Make sure a simple iterable can be wrapped as async and round trip
+    cleanly
+    """
+    lst = [1, 2, 3, 4]
+    async_round_trip = [x async for x in async_wrap_iter(lst)]
+    assert lst == async_round_trip
+
+
+@pytest.mark.asyncio
+async def test_async_wrap_iter_concurrent():
+    """Make sure that multiple wrapped synchronous iterators can interleve in
+    the event loop
+    """
+    lst1 = [1, 2, 3]
+    lst2 = [4, 5, 6]
+    iter1 = IterableHelper(lst1)
+    iter2 = IterableHelper(lst2)
+    aiter1 = async_wrap_iter(iter1)
+    aiter2 = async_wrap_iter(iter2)
+    zipped = [elt async for elt in azip(aiter1, aiter2)]
+    assert zipped == list(zip(lst1, lst2))
+
+
+@pytest.mark.asyncio
+async def test_async_wrap_iter_exception_propagation():
+    """Make sure exceptions thrown during iteration are raised in the parent
+    async context
+    """
+    with pytest.raises(RuntimeError):
+        async for _ in async_wrap_iter(IterableHelper([1, 2, 3], raise_element=1)):
+            pass
+
+
+@pytest.mark.asyncio
+async def test_async_wrap_iter_shared_executor():
+    """Make sure that a shared ThreadPoolExecutor can be used to manage the work
+    without exhausting thread counts
+    """
+    lst = [1, 2, 3, 4]
+    pool = ThreadPoolExecutor(max_workers=1)
+
+    start_event = threading.Event()
+
+    def slow_task():
+        start_event.set()
+        time.sleep(0.1)
+
+    # Start the slow task and make sure it's started
+    slow_future = pool.submit(slow_task)
+    start_event.wait()
+
+    # Use the same pool to start the async iteration
+    agen = async_wrap_iter(lst, pool)
+
+    # Make sure the async iteration is blocked
+    assert pool._work_queue.qsize() == 1
+
+    # Wait for the slow task to finish
+    slow_future.result()
```

Comment:
instead of relying on timing, can we use a `finish_event`?

```
def slow_task():
    start_event.set()
    finish_event.wait()
...

assert pool._work_queue.qsize() == 1
finish_event.set()
slow_future.result()
```

Reply:
Yep, much cleaner. Thanks!